// @ariakit/core/cjs/select/select-store.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _AIHTEA5Dcjs = require('../__chunks/AIHTEA5D.cjs');
require('../__chunks/GHET3DES.cjs');
require('../__chunks/52LQV3SV.cjs');


var _4OD4ZFRGcjs = require('../__chunks/4OD4ZFRG.cjs');
require('../__chunks/4L3HCXJY.cjs');








var _7GWXP6CQcjs = require('../__chunks/7GWXP6CQ.cjs');


var _RCYVACJOcjs = require('../__chunks/RCYVACJO.cjs');
require('../__chunks/EXM2AW52.cjs');


var _GDZQUFNPcjs = require('../__chunks/GDZQUFNP.cjs');




var _72I2GWXFcjs = require('../__chunks/72I2GWXF.cjs');

// src/select/select-store.ts
function createSelectStore(_a = {}) {
  var _b = _a, {
    combobox
  } = _b, props = _72I2GWXFcjs.__objRest.call(void 0, _b, [
    "combobox"
  ]);
  const store = _7GWXP6CQcjs.mergeStore.call(void 0, 
    props.store,
    _7GWXP6CQcjs.omit.call(void 0, combobox, [
      "value",
      "items",
      "renderedItems",
      "baseElement",
      "arrowElement",
      "anchorElement",
      "contentElement",
      "popoverElement",
      "disclosureElement"
    ])
  );
  _7GWXP6CQcjs.throwOnConflictingProps.call(void 0, props, store);
  const syncState = store.getState();
  const composite = _4OD4ZFRGcjs.createCompositeStore.call(void 0, _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, props), {
    store,
    virtualFocus: _RCYVACJOcjs.defaultValue.call(void 0, 
      props.virtualFocus,
      syncState.virtualFocus,
      true
    ),
    includesBaseElement: _RCYVACJOcjs.defaultValue.call(void 0, 
      props.includesBaseElement,
      syncState.includesBaseElement,
      false
    ),
    activeId: _RCYVACJOcjs.defaultValue.call(void 0, 
      props.activeId,
      syncState.activeId,
      props.defaultActiveId,
      null
    ),
    orientation: _RCYVACJOcjs.defaultValue.call(void 0, 
      props.orientation,
      syncState.orientation,
      "vertical"
    )
  }));
  const popover = _AIHTEA5Dcjs.createPopoverStore.call(void 0, _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, props), {
    store,
    placement: _RCYVACJOcjs.defaultValue.call(void 0, 
      props.placement,
      syncState.placement,
      "bottom-start"
    )
  }));
  const initialValue = new String("");
  const initialState = _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, composite.getState()), popover.getState()), {
    value: _RCYVACJOcjs.defaultValue.call(void 0, 
      props.value,
      syncState.value,
      props.defaultValue,
      initialValue
    ),
    setValueOnMove: _RCYVACJOcjs.defaultValue.call(void 0, 
      props.setValueOnMove,
      syncState.setValueOnMove,
      false
    ),
    selectElement: _RCYVACJOcjs.defaultValue.call(void 0, syncState.selectElement, null),
    labelElement: _RCYVACJOcjs.defaultValue.call(void 0, syncState.labelElement, null)
  });
  const select = _7GWXP6CQcjs.createStore.call(void 0, initialState, composite, popover, store);
  _7GWXP6CQcjs.setup.call(void 0, 
    select,
    () => _7GWXP6CQcjs.sync.call(void 0, select, ["value", "items"], (state) => {
      if (state.value !== initialValue)
        return;
      if (!state.items.length)
        return;
      const item = state.items.find(
        (item2) => !item2.disabled && item2.value != null
      );
      if ((item == null ? void 0 : item.value) == null)
        return;
      select.setState("value", item.value);
    })
  );
  _7GWXP6CQcjs.setup.call(void 0, 
    select,
    () => _7GWXP6CQcjs.sync.call(void 0, select, ["mounted", "items", "value"], (state) => {
      if (combobox)
        return;
      if (state.mounted)
        return;
      const values = _GDZQUFNPcjs.toArray.call(void 0, state.value);
      const lastValue = values[values.length - 1];
      if (lastValue == null)
        return;
      const item = state.items.find(
        (item2) => !item2.disabled && item2.value === lastValue
      );
      if (!item)
        return;
      select.setState("activeId", item.id);
    })
  );
  _7GWXP6CQcjs.setup.call(void 0, 
    select,
    () => _7GWXP6CQcjs.batch.call(void 0, select, ["setValueOnMove", "moves"], (state) => {
      const { mounted, value, activeId } = select.getState();
      if (!state.setValueOnMove && mounted)
        return;
      if (Array.isArray(value))
        return;
      if (!state.moves)
        return;
      if (!activeId)
        return;
      const item = composite.item(activeId);
      if (!item || item.disabled || item.value == null)
        return;
      select.setState("value", item.value);
    })
  );
  return _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, composite), popover), select), {
    combobox,
    setValue: (value) => select.setState("value", value),
    setSelectElement: (element) => select.setState("selectElement", element),
    setLabelElement: (element) => select.setState("labelElement", element)
  });
}


exports.createSelectStore = createSelectStore;
