// @ariakit/core/cjs/combobox/combobox-store.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _AIHTEA5Dcjs = require('../__chunks/AIHTEA5D.cjs');
require('../__chunks/GHET3DES.cjs');
require('../__chunks/52LQV3SV.cjs');


var _4OD4ZFRGcjs = require('../__chunks/4OD4ZFRG.cjs');
require('../__chunks/4L3HCXJY.cjs');






var _7GWXP6CQcjs = require('../__chunks/7GWXP6CQ.cjs');


var _RCYVACJOcjs = require('../__chunks/RCYVACJO.cjs');



var _EIRIJC7Tcjs = require('../__chunks/EIRIJC7T.cjs');
require('../__chunks/EXM2AW52.cjs');
require('../__chunks/GDZQUFNP.cjs');



var _72I2GWXFcjs = require('../__chunks/72I2GWXF.cjs');

// src/combobox/combobox-store.ts
var isSafariOnMobile = _EIRIJC7Tcjs.isSafari.call(void 0, ) && _EIRIJC7Tcjs.isTouchDevice.call(void 0, );
function createComboboxStore(props = {}) {
  var _a;
  _7GWXP6CQcjs.throwOnConflictingProps.call(void 0, props, props.store);
  const syncState = (_a = props.store) == null ? void 0 : _a.getState();
  const activeId = _RCYVACJOcjs.defaultValue.call(void 0, 
    props.activeId,
    syncState == null ? void 0 : syncState.activeId,
    props.defaultActiveId,
    null
  );
  const composite = _4OD4ZFRGcjs.createCompositeStore.call(void 0, _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, props), {
    activeId,
    includesBaseElement: _RCYVACJOcjs.defaultValue.call(void 0, 
      props.includesBaseElement,
      syncState == null ? void 0 : syncState.includesBaseElement,
      true
    ),
    orientation: _RCYVACJOcjs.defaultValue.call(void 0, 
      props.orientation,
      syncState == null ? void 0 : syncState.orientation,
      "vertical"
    ),
    focusLoop: _RCYVACJOcjs.defaultValue.call(void 0, props.focusLoop, syncState == null ? void 0 : syncState.focusLoop, true),
    focusWrap: _RCYVACJOcjs.defaultValue.call(void 0, props.focusWrap, syncState == null ? void 0 : syncState.focusWrap, true),
    virtualFocus: _RCYVACJOcjs.defaultValue.call(void 0, 
      props.virtualFocus,
      syncState == null ? void 0 : syncState.virtualFocus,
      !isSafariOnMobile
    )
  }));
  const popover = _AIHTEA5Dcjs.createPopoverStore.call(void 0, _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, props), {
    placement: _RCYVACJOcjs.defaultValue.call(void 0, 
      props.placement,
      syncState == null ? void 0 : syncState.placement,
      "bottom-start"
    )
  }));
  const initialValue = _RCYVACJOcjs.defaultValue.call(void 0, 
    props.value,
    syncState == null ? void 0 : syncState.value,
    props.defaultValue,
    ""
  );
  const initialState = _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, composite.getState()), popover.getState()), {
    value: initialValue,
    resetValueOnHide: _RCYVACJOcjs.defaultValue.call(void 0, 
      props.resetValueOnHide,
      syncState == null ? void 0 : syncState.resetValueOnHide,
      false
    ),
    activeValue: syncState == null ? void 0 : syncState.activeValue
  });
  const combobox = _7GWXP6CQcjs.createStore.call(void 0, initialState, composite, popover, props.store);
  _7GWXP6CQcjs.setup.call(void 0, 
    combobox,
    () => _7GWXP6CQcjs.sync.call(void 0, combobox, ["resetValueOnHide", "mounted"], (state) => {
      if (!state.resetValueOnHide)
        return;
      if (state.mounted)
        return;
      combobox.setState("value", initialValue);
    })
  );
  _7GWXP6CQcjs.setup.call(void 0, 
    combobox,
    () => _7GWXP6CQcjs.batch.call(void 0, combobox, ["mounted"], (state) => {
      if (state.mounted)
        return;
      combobox.setState("activeId", activeId);
      combobox.setState("moves", 0);
    })
  );
  _7GWXP6CQcjs.setup.call(void 0, 
    combobox,
    () => _7GWXP6CQcjs.sync.call(void 0, combobox, ["moves", "activeId"], (state, prevState) => {
      if (state.moves === prevState.moves) {
        combobox.setState("activeValue", void 0);
      }
    })
  );
  _7GWXP6CQcjs.setup.call(void 0, 
    combobox,
    () => _7GWXP6CQcjs.batch.call(void 0, combobox, ["moves", "renderedItems"], (state, prev) => {
      if (state.moves === prev.moves)
        return;
      const { activeId: activeId2 } = combobox.getState();
      const activeItem = composite.item(activeId2);
      combobox.setState("activeValue", activeItem == null ? void 0 : activeItem.value);
    })
  );
  return _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, popover), composite), combobox), {
    setValue: (value) => combobox.setState("value", value)
  });
}


exports.createComboboxStore = createComboboxStore;
