// @wordpress/block-editor/build/components/inserter/block-patterns-explorer/patterns-list.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _element = require("@wordpress/element");
var _i18n = require("@wordpress/i18n");
var _compose = require("@wordpress/compose");
var _components = require("@wordpress/components");
var _a11y = require("@wordpress/a11y");
var _blockPatternsList = _interopRequireDefault(require("../../block-patterns-list"));
var _useInsertionPoint = _interopRequireDefault(require("../hooks/use-insertion-point"));
var _usePatternsState = _interopRequireDefault(require("../hooks/use-patterns-state"));
var _inserterListbox = _interopRequireDefault(require("../../inserter-listbox"));
var _searchItems = require("../search-items");
var _blockPatternsPaging = _interopRequireDefault(require("../../block-patterns-paging"));
var _usePatternsPaging = _interopRequireDefault(require("../hooks/use-patterns-paging"));
var _blockPatternsTab = require("../block-patterns-tab");
var _blockPatternsSyncFilter = require("../block-patterns-sync-filter");
var _blockPatternsSourceFilter = require("../block-patterns-source-filter");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function PatternsListHeader({
  filterValue,
  filteredBlockPatternsLength,
  selectedCategory,
  patternCategories
}) {
  if (!filterValue) {
    return null;
  }
  let filter = filterValue;
  if (selectedCategory !== _blockPatternsTab.allPatternsCategory.name) {
    const category = patternCategories.find(patternCategory => patternCategory.name === selectedCategory);
    if (category) {
      filter = `${filter} - ${category?.label}`;
    }
  }
  return (0, _element.createElement)(_components.__experimentalHeading, {
    level: 2,
    lineHeight: '48px',
    className: "block-editor-block-patterns-explorer__search-results-count"
  }, (0, _i18n.sprintf)( /* translators: %d: number of patterns. %s: block pattern search query */
  (0, _i18n._n)('%1$d pattern found for "%2$s"', '%1$d patterns found for "%2$s"', filteredBlockPatternsLength), filteredBlockPatternsLength, filter));
}
function PatternList({
  searchValue,
  patternSourceFilter,
  selectedCategory,
  patternCategories
}) {
  const [patternSyncFilter, setPatternSyncFilter] = (0, _element.useState)('all');
  const container = (0, _element.useRef)();
  const debouncedSpeak = (0, _compose.useDebounce)(_a11y.speak, 500);
  const [destinationRootClientId, onInsertBlocks] = (0, _useInsertionPoint.default)({
    shouldFocusBlock: true
  });
  const [patterns,, onClickPattern] = (0, _usePatternsState.default)(onInsertBlocks, destinationRootClientId);
  const registeredPatternCategories = (0, _element.useMemo)(() => patternCategories.map(patternCategory => patternCategory.name), [patternCategories]);
  const filteredBlockPatterns = (0, _element.useMemo)(() => {
    const filteredPatterns = patterns.filter(pattern => {
      if ((0, _blockPatternsTab.isPatternFiltered)(pattern, patternSourceFilter, patternSyncFilter)) {
        return false;
      }
      if (selectedCategory === _blockPatternsTab.allPatternsCategory.name) {
        return true;
      }
      if (selectedCategory === 'uncategorized') {
        const hasKnownCategory = pattern.categories.some(category => registeredPatternCategories.includes(category));
        return !pattern.categories?.length || !hasKnownCategory;
      }
      return pattern.categories?.includes(selectedCategory);
    });
    if (!searchValue) {
      return filteredPatterns;
    }
    return (0, _searchItems.searchItems)(filteredPatterns, searchValue, patternSourceFilter);
  }, [searchValue, patternSourceFilter, patterns, selectedCategory, registeredPatternCategories, patternSyncFilter]);

  // Announce search results on change.
  (0, _element.useEffect)(() => {
    if (!searchValue) {
      return;
    }
    const count = filteredBlockPatterns.length;
    const resultsFoundMessage = (0, _i18n.sprintf)( /* translators: %d: number of results. */
    (0, _i18n._n)('%d result found.', '%d results found.', count), count);
    debouncedSpeak(resultsFoundMessage);
  }, [searchValue, debouncedSpeak, filteredBlockPatterns.length]);
  const pagingProps = (0, _usePatternsPaging.default)(filteredBlockPatterns, selectedCategory, container, patternSourceFilter);
  const hasItems = !!filteredBlockPatterns?.length;
  return (0, _element.createElement)("div", {
    className: "block-editor-block-patterns-explorer__list",
    ref: container
  }, (0, _element.createElement)(PatternsListHeader, {
    filterValue: searchValue || _blockPatternsSourceFilter.PATTERN_SOURCE_FILTERS[patternSourceFilter],
    filteredBlockPatternsLength: filteredBlockPatterns.length,
    selectedCategory: selectedCategory,
    patternCategories: patternCategories
  }), (0, _element.createElement)(_inserterListbox.default, null, patternSourceFilter === _blockPatternsSourceFilter.PATTERN_TYPES.user && !searchValue && (0, _element.createElement)(_blockPatternsSyncFilter.BlockPatternsSyncFilter, {
    patternSyncFilter: patternSyncFilter,
    setPatternSyncFilter: setPatternSyncFilter
  }), hasItems && (0, _element.createElement)(_blockPatternsList.default, {
    shownPatterns: pagingProps.categoryPatternsAsyncList,
    blockPatterns: pagingProps.categoryPatterns,
    onClickPattern: onClickPattern,
    isDraggable: false
  }), pagingProps.numPages > 1 && (0, _element.createElement)(_blockPatternsPaging.default, {
    ...pagingProps
  })));
}
var _default = PatternList;
exports.default = _default;
