// @wordpress/block-editor/build/components/inserter/block-patterns-explorer/sidebar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _element = require("@wordpress/element");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _blockPatternsSourceFilter = _interopRequireDefault(require("../block-patterns-source-filter"));
var _blockPatternsTab = require("../block-patterns-tab");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function PatternCategoriesList({
  selectedCategory,
  patternCategories,
  onClickCategory
}) {
  const baseClassName = 'block-editor-block-patterns-explorer__sidebar';
  return (0, _element.createElement)("div", {
    className: `${baseClassName}__categories-list`
  }, patternCategories.map(({
    name,
    label
  }) => {
    return (0, _element.createElement)(_components.Button, {
      key: name,
      label: label,
      className: `${baseClassName}__categories-list__item`,
      isPressed: selectedCategory === name,
      onClick: () => {
        onClickCategory(name);
      }
    }, label);
  }));
}
function PatternsExplorerSearch({
  searchValue,
  setSearchValue
}) {
  const baseClassName = 'block-editor-block-patterns-explorer__search';
  return (0, _element.createElement)("div", {
    className: baseClassName
  }, (0, _element.createElement)(_components.SearchControl, {
    __nextHasNoMarginBottom: true,
    onChange: setSearchValue,
    value: searchValue,
    label: (0, _i18n.__)('Search for patterns'),
    placeholder: (0, _i18n.__)('Search')
  }));
}
function PatternExplorerSidebar({
  selectedCategory,
  patternCategories,
  onClickCategory,
  patternSourceFilter,
  setPatternSourceFilter,
  searchValue,
  setSearchValue
}) {
  const baseClassName = 'block-editor-block-patterns-explorer__sidebar';
  return (0, _element.createElement)("div", {
    className: baseClassName
  }, (0, _element.createElement)(PatternsExplorerSearch, {
    searchValue: searchValue,
    setSearchValue: setSearchValue
  }), (0, _element.createElement)(_blockPatternsSourceFilter.default, {
    value: patternSourceFilter,
    onChange: value => {
      setPatternSourceFilter(value);
      onClickCategory(_blockPatternsTab.allPatternsCategory.name);
    }
  }), !searchValue && (0, _element.createElement)(PatternCategoriesList, {
    selectedCategory: selectedCategory,
    patternCategories: patternCategories,
    onClickCategory: onClickCategory
  }));
}
var _default = PatternExplorerSidebar;
exports.default = _default;
