// @wordpress/block-editor/build/components/inserter/block-patterns-sync-filter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BlockPatternsSyncFilter = BlockPatternsSyncFilter;
exports.SYNC_TYPES = void 0;
var _element = require("@wordpress/element");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
const SYNC_TYPES = {
  full: 'fully',
  unsynced: 'unsynced'
};
exports.SYNC_TYPES = SYNC_TYPES;
const patternSyncOptions = [{
  value: 'all',
  label: (0, _i18n.__)('All')
}, {
  value: SYNC_TYPES.full,
  label: (0, _i18n.__)('Synced')
}, {
  value: SYNC_TYPES.unsynced,
  label: (0, _i18n.__)('Standard')
}];
function BlockPatternsSyncFilter({
  setPatternSyncFilter,
  patternSyncFilter
}) {
  return (0, _element.createElement)(_components.SelectControl, {
    className: "block-editor-patterns__sync-status-filter",
    label: (0, _i18n.__)('Syncing'),
    options: patternSyncOptions,
    value: patternSyncFilter,
    onChange: setPatternSyncFilter,
    "aria-label": (0, _i18n.__)('Filter patterns by sync type')
  });
}
